File: pfms-ros/pfms/scripts/pose_to_tf.py
#!/usr/bin/env python3
"""
Simple node to convert Gazebo pose to TF with frame prefix
"""
import rclpy
from rclpy.node import Node
from rclpy.qos import DurabilityPolicy
from rclpy.qos import HistoryPolicy
from rclpy.qos import QoSProfile
from rclpy.qos import ReliabilityPolicy
from geometry_msgs.msg import PoseArray
from nav_msgs.msg import Odometry
from tf2_ros import TransformBroadcaster
from geometry_msgs.msg import TransformStamped
import time


class PoseToTF(Node):
    def __init__(self):
        super().__init__('pose_to_tf')
        
        # Declare parameters
        self.declare_parameter('frame_prefix', '')
        self.declare_parameter('parent_frame', 'world')
        self.declare_parameter('child_frame', 'base_footprint')
        self.declare_parameter('publish_rate_hz', 20.0)
        self.declare_parameter('input_mode', 'pose_array')
        
        # Get parameters
        self.frame_prefix = self.get_parameter('frame_prefix').value
        self.parent_frame = self.get_parameter('parent_frame').value
        child_frame_base = self.get_parameter('child_frame').value
        self.child_frame = f"{self.frame_prefix}{child_frame_base}"
        publish_rate_raw = self.get_parameter('publish_rate_hz').value
        publish_rate_hz = float(publish_rate_raw) if publish_rate_raw is not None else 20.0
        
        # Protect against invalid rates
        if publish_rate_hz <= 0.0:
            publish_rate_hz = 20.0
            
        input_mode_raw = self.get_parameter('input_mode').value
        input_mode = str(input_mode_raw) if input_mode_raw is not None else 'pose_array'

        # Create TF broadcaster
        self.tf_broadcaster = TransformBroadcaster(self)
        
        # Subscribe with sensor-data style QoS to reduce transport overhead.
        qos = QoSProfile(
            reliability=ReliabilityPolicy.BEST_EFFORT,
            durability=DurabilityPolicy.VOLATILE,
            history=HistoryPolicy.KEEP_LAST,
            depth=1,
        )
    
        # Create a timer instead of relying on message callbacks
        # This decouples processing rate from message rate
        self.latest_pose = None
        self.timer = self.create_timer(1.0 / publish_rate_hz, self.timer_callback)

        # Profiling metrics
        self.declare_parameter('enable_profiling', False)
        self.enable_profiling = self.get_parameter('enable_profiling').value
        self.callback_count = 0
        self.timer_count = 0
        self.callback_time_total = 0.0
        self.timer_time_total = 0.0
        self.last_profile_report = time.time()
        
        if self.enable_profiling:
            self.profile_timer = self.create_timer(5.0, self.report_profiling)
            self.get_logger().info('Profiling enabled - reports every 5 seconds')

        if input_mode == 'odom':
            self.source_sub = self.create_subscription(
                Odometry,
                'odom',
                self.odom_store_callback,
                qos
            )
        else:
            self.source_sub = self.create_subscription(
                PoseArray,
                'audibot/pose',
                self.pose_store_callback,
                qos
            )
        
        self.get_logger().info(
            f'Publishing TF: {self.parent_frame} -> {self.child_frame} at {publish_rate_hz:.1f} Hz (mode={input_mode})'
        )
    
    def pose_store_callback(self, msg):
        """Store latest pose without processing - called at high frequency"""
        if self.enable_profiling:
            start = time.perf_counter()
        
        if len(msg.poses) > 0:
            self.latest_pose = msg.poses[0]
        
        if self.enable_profiling:
            self.callback_time_total += (time.perf_counter() - start)
            self.callback_count += 1

    def odom_store_callback(self, msg):
        """Store latest pose from odometry without processing"""
        if self.enable_profiling:
            start = time.perf_counter()
        
        self.latest_pose = msg.pose.pose
        
        if self.enable_profiling:
            self.callback_time_total += (time.perf_counter() - start)
            self.callback_count += 1

    def timer_callback(self):
        """Publish TF at fixed rate from latest stored pose"""
        if self.enable_profiling:
            start = time.perf_counter()
        
        if self.latest_pose is None:
            if self.enable_profiling:
                self.timer_time_total += (time.perf_counter() - start)
                self.timer_count += 1
            return
        
        pose = self.latest_pose
        
        # Create and publish transform
        t = TransformStamped()
        t.header.stamp = self.get_clock().now().to_msg()
        t.header.frame_id = self.parent_frame
        t.child_frame_id = self.child_frame
        
        t.transform.translation.x = pose.position.x
        t.transform.translation.y = pose.position.y
        t.transform.translation.z = pose.position.z
        
        t.transform.rotation.x = pose.orientation.x
        t.transform.rotation.y = pose.orientation.y
        t.transform.rotation.z = pose.orientation.z
        t.transform.rotation.w = pose.orientation.w
        
        self.tf_broadcaster.sendTransform(t)
        
        if self.enable_profiling:
            self.timer_time_total += (time.perf_counter() - start)
            self.timer_count += 1

    def report_profiling(self):
        """Report profiling statistics every 5 seconds"""
        elapsed = time.time() - self.last_profile_report
        
        if self.callback_count > 0:
            avg_callback_us = (self.callback_time_total / self.callback_count) * 1e6
            callback_hz = self.callback_count / elapsed
        else:
            avg_callback_us = 0.0
            callback_hz = 0.0
        
        if self.timer_count > 0:
            avg_timer_us = (self.timer_time_total / self.timer_count) * 1e6
            timer_hz = self.timer_count / elapsed
        else:
            avg_timer_us = 0.0
            timer_hz = 0.0
        
        total_time_ms = (self.callback_time_total + self.timer_time_total) * 1000
        cpu_percent = (total_time_ms / (elapsed * 1000)) * 100
        
        self.get_logger().info(
            f'\n=== Profiling Report (last {elapsed:.1f}s) ===\n'
            f'Callbacks: {self.callback_count} calls ({callback_hz:.1f} Hz), avg {avg_callback_us:.2f} μs/call\n'
            f'Timer: {self.timer_count} calls ({timer_hz:.1f} Hz), avg {avg_timer_us:.2f} μs/call\n'
            f'Total compute: {total_time_ms:.2f} ms, est CPU: {cpu_percent:.1f}%'
        )
        
        # Reset counters
        self.callback_count = 0
        self.timer_count = 0
        self.callback_time_total = 0.0
        self.timer_time_total = 0.0
        self.last_profile_report = time.time()


def main(args=None):
    rclpy.init(args=args)
    node = PoseToTF()
    rclpy.spin(node)
    node.destroy_node()
    rclpy.shutdown()


if __name__ == '__main__':
    main()

File: pfms-ros/pfms/src/pose_to_tf.cpp
/**
 * @brief Convert Gazebo pose to TF with frame prefix
 * 
 * This is a C++ implementation that is more CPU-efficient than the Python version.
 * It subscribes to either PoseArray or Odometry and publishes TF transforms
 * at a configurable fixed rate.
 */

#include <memory>
#include <string>
#include <chrono>

#include "rclcpp/rclcpp.hpp"
#include "geometry_msgs/msg/pose_array.hpp"
#include "geometry_msgs/msg/pose.hpp"
#include "nav_msgs/msg/odometry.hpp"
#include "tf2_ros/transform_broadcaster.h"
#include "geometry_msgs/msg/transform_stamped.hpp"

using namespace std::chrono_literals;

class PoseToTF : public rclcpp::Node
{
public:
  PoseToTF()
      : Node("pose_to_tf"),
        latest_pose_received_(false)
  {
    // Declare parameters
    this->declare_parameter<std::string>("frame_prefix", "");
    this->declare_parameter<std::string>("parent_frame", "world");
    this->declare_parameter<std::string>("child_frame", "base_footprint");
    this->declare_parameter<double>("publish_rate_hz", 20.0);
    this->declare_parameter<std::string>("input_mode", "pose_array");

    // Get parameters
    frame_prefix_ = this->get_parameter("frame_prefix").as_string();
    parent_frame_ = this->get_parameter("parent_frame").as_string();
    std::string child_frame_base = this->get_parameter("child_frame").as_string();
    child_frame_ = frame_prefix_ + child_frame_base;
    double publish_rate_hz = this->get_parameter("publish_rate_hz").as_double();
    input_mode_ = this->get_parameter("input_mode").as_string();

    // Protect against invalid rates
    if (publish_rate_hz <= 0.0) {
      publish_rate_hz = 20.0;
    }

    // Initialize TF broadcaster
    tf_broadcaster_ = std::make_unique<tf2_ros::TransformBroadcaster>(*this);

    // Set up QoS profile for sensor data (best effort, volatile)
    auto qos = rclcpp::QoS(rclcpp::KeepLast(1))
                   .reliability(RMW_QOS_POLICY_RELIABILITY_BEST_EFFORT)
                   .durability(RMW_QOS_POLICY_DURABILITY_VOLATILE);

    // Create subscription based on input mode
    if (input_mode_ == "odom") {
      odom_sub_ = this->create_subscription<nav_msgs::msg::Odometry>(
          "odom", qos,
          std::bind(&PoseToTF::odomCallback, this, std::placeholders::_1));
    } else {
      pose_array_sub_ = this->create_subscription<geometry_msgs::msg::PoseArray>(
          "audibot/pose", qos,
          std::bind(&PoseToTF::poseArrayCallback, this, std::placeholders::_1));
    }

    // Create timer for fixed-rate TF publishing (decouples from message rate)
    auto timer_period = std::chrono::duration<double>(1.0 / publish_rate_hz);
    timer_ = this->create_wall_timer(
        std::chrono::duration_cast<std::chrono::nanoseconds>(timer_period),
        std::bind(&PoseToTF::timerCallback, this));

    RCLCPP_INFO(this->get_logger(),
                "Publishing TF: %s -> %s at %.1f Hz (mode=%s)",
                parent_frame_.c_str(), child_frame_.c_str(),
                publish_rate_hz, input_mode_.c_str());
  }

private:
  void poseArrayCallback(const geometry_msgs::msg::PoseArray::SharedPtr msg)
  {
    if (!msg->poses.empty()) {
      // Store latest pose without processing - callback is lightweight
      latest_pose_ = msg->poses[0];
      latest_pose_received_ = true;
    }
  }

  void odomCallback(const nav_msgs::msg::Odometry::SharedPtr msg)
  {
    // Store latest pose from odometry
    latest_pose_ = msg->pose.pose;
    latest_pose_received_ = true;
  }

  void timerCallback()
  {
    if (!latest_pose_received_) {
      return;
    }

    // Create and publish transform
    geometry_msgs::msg::TransformStamped t;
    t.header.stamp = this->get_clock()->now();
    t.header.frame_id = parent_frame_;
    t.child_frame_id = child_frame_;

    t.transform.translation.x = latest_pose_.position.x;
    t.transform.translation.y = latest_pose_.position.y;
    t.transform.translation.z = latest_pose_.position.z;

    t.transform.rotation.x = latest_pose_.orientation.x;
    t.transform.rotation.y = latest_pose_.orientation.y;
    t.transform.rotation.z = latest_pose_.orientation.z;
    t.transform.rotation.w = latest_pose_.orientation.w;

    tf_broadcaster_->sendTransform(t);
  }

  // Subscriptions
  rclcpp::Subscription<geometry_msgs::msg::PoseArray>::SharedPtr pose_array_sub_;
  rclcpp::Subscription<nav_msgs::msg::Odometry>::SharedPtr odom_sub_;

  // Timer for fixed-rate publishing
  rclcpp::TimerBase::SharedPtr timer_;

  // TF broadcaster
  std::unique_ptr<tf2_ros::TransformBroadcaster> tf_broadcaster_;

  // Latest pose storage
  geometry_msgs::msg::Pose latest_pose_;
  bool latest_pose_received_;

  // Parameters
  std::string frame_prefix_;
  std::string parent_frame_;
  std::string child_frame_;
  std::string input_mode_;
};

int main(int argc, char** argv)
{
  rclcpp::init(argc, argv);
  rclcpp::spin(std::make_shared<PoseToTF>());
  rclcpp::shutdown();
  return 0;
}
